### PR TITLE
[`pipeline`] A simple fix for half-precision & 8bit models 

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Currently on the `main` branch of `transformers` if a user wants to run a `pipeline` using large models (thus, ideally loaded with `device_map=...`) and in half precision (or in int8), they may encounter some issues when calling `pipeline` with `top_p` & `top_k` sampling:
```bash
RuntimeError: "topk_cpu" not implemented for 'Half'
```

## Snippet to reproduce & explanations:

```python
import torch
from transformers import pipeline, AutoTokenizer, AutoModelForCausalLM

model_id = "EleutherAI/gpt-neo-125M"

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", torch_dtype=torch.float16)

pipe = pipeline("text-generation", model=model, tokenizer=tokenizer, max_length=20, temperature=1, do_sample=True, top_p=0.95, top_k=60, num_return_sequences=3)
text = "What can you tell me about the LHC?"
response = pipe(text)
print(response[0]["generated_text"])
```

This is because the `input_ids` are automatically set on `cpu` since the argument `device` is not passed when initializing the `pipeline`. A model that is loaded with `device_map=...` (i.e. with `accelerate`) always sets the output tensor of the model to the `device` of the input tensor thanks to the forward hooks. Therefore when calling the top_k method, the output tensor is in fp16 (because the model has been loaded in fp16) & on `cpu` hence the torch error above.

Currently a hack to fix this is to add `device=0` when initializing the `pipeline` but this leads to inconsistent and undesirable behaviours for some cases, for example when loading large models in several GPUs, since the call `model.to(self.device)` will break some internals (the hooks will be still there but the weights will be set on the wrong devices). A snippet to reproduce below:

```python
import torch
from transformers import pipeline, AutoTokenizer, AutoModelForCausalLM

model_id = "EleutherAI/gpt-neo-125M"

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id, device_map="balanced", torch_dtype=torch.float16)

pipe = pipeline("text-generation", model=model, tokenizer=tokenizer, max_length=20, temperature=1, do_sample=True, top_p=0.95, top_k=60, num_return_sequences=3, device=0)
text = "What can you tell me about the LHC?"
response = pipe(text)
print(response[0]["generated_text"])
```

adding this hack also breaks the usage of `pipeline` with int8 models, since the `to` method is blocked for these models:

```bash
ValueError: `.to` is not supported for `8-bit` models. Please use the model as it is, since the model has already been set to the correct devices and casted to the correct `dtype`.
```

Thus, I propose to fix this by simply checking whether a model has been loaded with `accelerate` by looking at the attribute `hf_device_map` , and set the model on the correct device only if it has not been loaded with accelerate as backend. This fixes 3 bugs: using `pipeline` with a fp16 model that has been loaded with `accelerate` without having any error in case of multi-gpu usage, using `pipeline` with a fp16 model w `accelerate` & sampling strategies, and using `pipeline` with int8 models & sampling strategies.

cc @sgugger @Narsil

### CODE REVIEW

**Comment by Narsil** [comment]
Thanks for the well thought out issue and proposed fix. 

I don't particularly like the fix because it depends on some weird internal and still forces users to use `device_map` and `device` iiuc.

Couldn't we just use `device_map` and use `accelerate` api to figure out where to put the inputs? (most likely cuda:0 but still cpu if no gpus are available I think.)
That or just do something special for `device_map` without asking where the model is (if the API doesn't exist or is tricky).

Imo using `device_map` and `device` should be an error (ambiguous intent)

**Comment by younesbelkada** [comment]
Thanks for the feedback @Narsil !
I think 
> Imo using device_map and device should be an error (ambiguous intent)
Makes sense !
Another fix would be to force-upcast the logits in fp32 when doing top_k & top_p sampling on the generation side only if the logits are on `cpu`, is this solution a reasonable fix @gante ? Happy to open a PR to fix it!

**Comment by Narsil** [comment]
> force-upcast 

I would highly advise against it too. There's a limit to magic. Doing half precision on cpu should crash in a lot of places. We shouldn't upcast on behalf of a user that explicitely asked for half precision imo. That's breaking user intent.
But the user also asked for GPU, that's where we're breaking his intent and that's what should be fixed IMO.

Does `accelerate` allow to know on which device is the start of the model ?

**Comment by younesbelkada** [comment]
I see, makes sense! 

> Does accelerate allow to know on which device is the start of the model ?

I am not sure here, maybe @sgugger & @muellerzr knows better

**Comment by Narsil** [comment]
> I am not sure here, maybe @sgugger & @muellerzr knows better

if not the pipeline could have the simplest heuristic `'cuda:0' if torch.cuda.is_avalaible() else 'cpu'` which should work 99% of the time. 
But it wouldn't if a user specified an odd map (which is why having direct access would be better).

**Comment by Narsil** [comment]
Sure. Let's update the doc too.

**Comment by sgugger** [comment]
Side notes:
- you should probably do something if the user passes a device and the model has a `hf_device_map` (at least a warning) as the line `model.to(device)` will probably screw things up (it will at least error if there are some weights offloaded)
- the device on which the model is executed is determined by this rule in Accelerate, maybe you should use the same code? I can also store it on the Accelerate side in a special attribute but then you'd have to wait for a release.
```py
    if set(device_map.values()) == {"cpu"} or set(device_map.values()) == {"cpu", "disk"}:
        main_device = "cpu"
    else:
        main_device = [d for d in device_map.values() if d not in ["cpu", "disk"]][0]
```

**Comment by younesbelkada** [comment]
Thanks a lot for the valuable feedback @Narsil @sgugger ! 
I updated the PR and added more clarification (and also a new section) on the docs

File: src/transformers/pipelines/base.py [hunk 1]
```diff
@@ -769,18 +769,41 @@ def __init__(
...
-            self.device = device
+            self.device = device if device is not None else -1
         self.torch_dtype = torch_dtype
         self.binary_output = binary_output
 
         # Special handling
-        if self.framework == "pt" and self.device.type != "cpu":
-            self.model = self.model.to(self.device)
+        if self.framework == "pt" and device is not None:
+            self.model = self.model.to(device=self.device)
```
Comment (Narsil) [inline src/transformers/pipelines/base.py:783]: Don't mix `self.device` and `device` . This is super error prone.

The proposed change I made was at least explicit about it's default value.
I really think this needs to be changed. Too many opportunities to introduce bugs later on.

- Set the default value (if no value provided)
- Handle `device` to create `self.device`.
- Use `self.device` everywhere.

File: docs/source/en/pipeline_tutorial.mdx [hunk 2]
```diff
@@ -257,4 +259,37 @@ sudo apt install -y tesseract-ocr
...
+
+```py
+# pip install accelerate bitsandbytes
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline
+
+model = AutoModelForCausalLM.from_pretrained("bigscience/bloom", device_map="auto", load_in_8bit=True)
+tokenizer = AutoTokenizer.from_pretrained("bigscience/bloom")
+
+pipe = pipeline("text-generation", model=model, tokenizer=tokenizer)
```
Comment (Narsil) [inline docs/source/en/pipeline_tutorial.mdx:293]:
```suggestion
import torch
from transformers import pipeline

pipe = pipeline(model="bigscience/bloom", device_map="auto", model_kwargs={"load_in_8bit":True})
```

**Review by Narsil — COMMENTED** [review]
Thanks for the update.

I propose to simplify the examples a lot.

Then, regarding the actual code, I think your proposed change is entangling and confusing `device`/`self.device` more than it should. At least more than my original modification.

We could go even further in "cleanliness" and handle all that logic within `pipeline` function and not in `Pipeline.__init__` since at that point the model should be a black box (that includes moving the model on device). 

Modifying the model on device things **will** break calls like `TranslationPipeline(model=MyModel(), device=0)`. I don't think we can realistically break that, but we can probably figure out a non breaking change that does handle `device_map` better.

File: docs/source/en/pipeline_tutorial.mdx [hunk 3]
```diff
@@ -257,4 +259,37 @@ sudo apt install -y tesseract-ocr
...
+Let's assume you fullfill the hardware requirements to run a large model such as `bloom` (which has 176B parameters, so ~350GB in `bfloat16`). First load your model
+using `device_map="auto"`
+
+```py
+# pip install accelerate
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline
+
+model = AutoModelForCausalLM.from_pretrained("bigscience/bloom", torch_dtype=torch.bfloat16, device_map="auto")
+tokenizer = AutoTokenizer.from_pretrained("bigscience/bloom")
```
Comment (sgugger) [inline docs/source/en/pipeline_tutorial.mdx:277]: Maybe use a smaller example and say in a note the user can replace it by BLOOM?

**Review by sgugger — COMMENTED** [review]
Agreed with all of @Narsil 's comments!

File: src/transformers/pipelines/base.py [hunk 4]
```diff
@@ -769,18 +769,41 @@ def __init__(
...
+
+            hf_device_map = getattr(self.model, "hf_device_map", None)
+            if hf_device_map is not None:
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/pipelines/base.py:786]: 💯 on this @Narsil

File: src/transformers/pipelines/base.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/pipelines/base.py:783]: Yeah I somehow didn't fully considered your proposition in https://github.com/huggingface/transformers/pull/21479#issuecomment-1420747654 - I think it's wiser to revert my changes with yours!

File: docs/source/en/pipeline_tutorial.mdx (hunk 3 above)
Comment (younesbelkada) [reply docs/source/en/pipeline_tutorial.mdx]: Sure!

**Review by sgugger — APPROVED** [review]
Thanks for iterating!

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 5]
```diff
@@ -287,9 +287,9 @@ class AutomaticSpeechRecognitionPipeline(ChunkPipeline):
             installed. If no framework is specified, will default to the one currently installed. If no framework is
             specified and both frameworks are installed, will default to the framework of the `model`, or to PyTorch if
             no model is provided.
-        device (`int`, *optional*, defaults to -1):
-            Device ordinal for CPU/GPU supports. Setting this to -1 will leverage CPU, a positive will run the model on
-            the associated CUDA device id.
+        device (Union[`int`, `torch.device`], *optional*, defaults to `None`):
```
Comment (sgugger) [inline src/transformers/pipelines/automatic_speech_recognition.py:290]:
```suggestion
        device (Union[`int`, `torch.device`], *optional*):
```
Don't write the default if it's None ;-)

Comment (younesbelkada) [reply src/transformers/pipelines/automatic_speech_recognition.py:290]: Ah yes! :D

**Review by Narsil — APPROVED** [review]
LGTM.

**Comment by younesbelkada** [comment]
the failing test is indenpendent to our PR! Merging!
Thanks for all your comments!

**Pull request merged**